dvc/repo/artifacts.py

```diff
@@ -0,0 +1,44 @@
+import os
+from typing import TYPE_CHECKING, Any, Dict
+
+from dvc.annotations import Artifact
+from dvc.dvcfile import FileMixin
+from dvc.utils import relpath
+
+if TYPE_CHECKING:
+    from dvc.repo import Repo
+
+
+class ArtifactsFile(FileMixin):
+    from dvc.schema import ARTIFACTS_SCHEMA as SCHEMA
+
+    def dump(self, stage, **kwargs):
+        raise NotImplementedError
+
+    def merge(self, ancestor, other, allowed=None):
+        raise NotImplementedError
+
+
+class Artifacts:
+    repo: "Repo"
+
+    def __init__(self, repo) -> None:
+        self.repo = repo
+
+    def read(self):
+        artifacts: Dict[str, Dict[str, Any]] = {}
+        for (
+            dvcfile,
+            dvcfile_artifacts,
+        ) in self.repo.index._artifacts.items():  # pylint: disable=protected-access
+            # read the artifacts.yaml file if needed
+            if isinstance(dvcfile_artifacts, str):
+                dvcfile_artifacts = ArtifactsFile(
+                    self.repo,
+                    os.path.join(os.path.dirname(dvcfile), dvcfile_artifacts),
+                    verify=False,
+                ).load()
+            artifacts[relpath(dvcfile, self.repo.root_dir)] = {
+                name: Artifact(**value) for name, value in dvcfile_artifacts.items()
```

Comment:
Ok, another question that came to my mind. In GTO we limited possible `name`s (it's also called `ID` in DVC): only `a-Z-_` chars were allowed.  

This was made for simplicity - it can be problematic to have Git tags with chars like `?;,.`, etc.
We also have a specific usage of symbols `#!` in Git tag names, so we need to reserve them as well to avoid ambiguity.

Now it also allows us to introduce proper [handling `dvc.yaml`s in subdirs](https://github.com/iterative/gto/issues/198#issuecomment-1480901662) (e.g. `dvclive/dvc.yaml`) in non-ambiguous way.

So I guess we need to validate names in DVC the same way we did in GTO. Now 2 questions:

1. Do we do this with anything now? I tried arbitrary string as plot ID and it worked. Also I don't see any validation in [dvcyaml-schema](https://github.com/iterative/dvcyaml-schema/blob/master/schema.json) (maybe not best place to search for the answer though)

2. How we handle the case when `ID` (`name`) was written incorrectly? Do we issue a warning and ignore the artifact?

WDYT @dberenbaum @daavoo @skshetry ?

Reply:
> Do we do this with anything now? I tried arbitrary string as plot ID and it worked. Also I don't see any validation in [dvcyaml-schema](https://github.com/iterative/dvcyaml-schema/blob/master/schema.json) (maybe not best place to search for the answer though)

I think the closest thing is how we check the provided name for an experiment:

https://github.com/iterative/dvc/blob/efc2d7ad3d2cedcd3d13dc8ddc767751a7819431/dvc/repo/experiments/__init__.py#L273

https://github.com/iterative/dvc/blob/efc2d7ad3d2cedcd3d13dc8ddc767751a7819431/dvc/repo/experiments/utils.py#L280-L286

> How we handle the case when ID (name) was written incorrectly? Do we issue a warning and ignore the artifact?

In metrics/params/plots collection there is the error handler where instead of failing the message is returned so UIs (VSCode / Studio) can display it.
Maybe it makes sense to do the same here. Who would use the `artifacts` and for what?

Reply:
I guess the usual use will be:

1. Add `artifacts: myart` to `dvc.yaml`
2. Create git tags like `myart@v0.0.1` that start CI
3. Get artifact in CI to deploy/publish it.

So it would be good if we can show some feedback in VScode/Studio, but I think in VSCode it should come from some checks in dvcyaml-schema.

Reply:
> Ok, another question that came to my mind. In GTO we limited possible names (it's also called ID in DVC): only a-Z-_ chars were allowed.

In the contemporary world, it does not make sense to limit to only ascii. We try to support whenever we can.

Reply:
>  it does not make sense to limit to only ascii. We try to support whenever we can.

@skshetry it is a limitation on `git` tags itself, not sure if we can bypass that (as the artifact id is supposed to be linked to a valid git tag)

Reply:
Sorry, I quoted the wrong thing. Was trying to answer the first question.